cmd/habitat/community/manager.go

```diff
@@ -1,47 +1,92 @@
 package community
 
 import (
+	"encoding/base64"
+	"encoding/json"
+	"errors"
 	"fmt"
+	"io/ioutil"
 	"os"
 	"path"
 	"path/filepath"
+	"sync"
+	"time"
 
 	"github.com/eagraf/habitat/cmd/habitat/community/consensus/cluster"
+	"github.com/eagraf/habitat/cmd/habitat/community/state"
+	"github.com/eagraf/habitat/cmd/habitat/procs"
 	"github.com/eagraf/habitat/cmd/habitat/proxy"
 	"github.com/eagraf/habitat/pkg/compass"
 	"github.com/eagraf/habitat/pkg/ipfs"
 	"github.com/eagraf/habitat/structs/community"
 	"github.com/google/uuid"
 	"github.com/libp2p/go-libp2p-core/host"
+	"github.com/rs/zerolog/log"
 )
 
 type Manager struct {
-	Path    string
-	config  *ipfs.IPFSConfig
-	p2pHost host.Host
+	Path           string
+	config         *ipfs.IPFSConfig
+	p2pHost        host.Host
+	processManager procs.Manager
 
-	clusterManager *cluster.ClusterManager
-	communities    []*community.Community
+	clusterManager  *cluster.ClusterManager
+	communities     map[string]*state.CommunityStateMachine
+	communitiesLock *sync.Mutex
 }
 
-func NewManager(path string, proxyRules *proxy.RuleSet, host host.Host) (*Manager, error) {
+func NewManager(path string, procManager *procs.Manager, proxyRules *proxy.RuleSet, host host.Host) (*Manager, error) {
 	clusterManager := cluster.NewClusterManager(host)
 
 	err := clusterManager.Start(proxyRules)
 	if err != nil {
 		return nil, fmt.Errorf("error starting cluster manager: %s", err)
 	}
 
-	return &Manager{
+	manager := &Manager{
 		Path: path,
 		config: &ipfs.IPFSConfig{
 			CommunitiesPath: path,
 			// TODO: @arushibandi remove this usage of compass
 			StartCmd: filepath.Join(compass.ProcsPath(), "bin", "amd64-darwin", "start-ipfs"),
 		},
-		p2pHost:        host,
-		clusterManager: clusterManager,
-	}, nil
+		p2pHost:         host,
+		processManager:  *procManager,
+		clusterManager:  clusterManager,
+		communities:     make(map[string]*state.CommunityStateMachine),
+		communitiesLock: &sync.Mutex{},
+	}
+
+	// Restart any existing communities
+	comDirs, err := ioutil.ReadDir(compass.CommunitiesPath())
+	if err == nil {
+		for _, dir := range comDirs {
+			updateChan, err := clusterManager.RestoreNode(dir.Name())
+			if err != nil {
+				log.Error().Err(err).Msgf("error restoring cluster for community %s", dir.Name())
+			}
+
+			initState := &community.CommunityState{
+				CommunityID: dir.Name(),
+			}
+			stateMachine, err := state.NewCommunityStateMachine(initState,
+				updateChan, &ClusterDispatcher{
```

Comment:
i think i need to understand this structure of setting up the community state machine a little better. why are cluster dispatcher and community executor different things? and why is updateChan also separate from these two?

Reply:
The dispatcher is responsible for sending new transitions to the network, while executor is responsible for figuring out what to do after a commited state transition is received. Since all of the raft stuff is sitting in between them, it makes sense for them to be separate (I tried different ways of unifying them earlier, but I think it makes the most sense to keep them separated. `updateChan` is what sends the transitions from the raft instance to the executor